src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -178,6 +179,7 @@ protected internal void AddGas(long gas)
         protected override void OnFault(Exception ex)
         {
             FaultException = ex;
+            notifications = null;
```

Comment:
Contract won't be able to know it because the transaction is failed. ApplicationLogs can receive all notifications from events. So we don't need to keep it.